storage/range.go

```diff
@@ -789,32 +786,71 @@ func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.I
 // underlying state machine (i.e. the engine).
 // When certain critical operations fail, a replicaCorruptionError may be
 // returned and must be handled by the caller.
-func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNode proto.RaftNodeID, args proto.Request, reply proto.Response) (rErr error) {
+func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNode proto.RaftNodeID, args proto.Request, reply proto.Response) error {
 	if index <= 0 {
 		log.Fatalc(ctx, "raft command index is <= 0")
 	}
 
-	committed := false
-	// The very last thing we do before returning is move the applied index
-	// forward, unless that has already happened as part of a successfully
-	// committed batch.
-	defer func() {
-		if !committed {
-			// We didn't commit the batch, but advance the last applied index nonetheless.
-			if err := setAppliedIndex(r.rm.Engine(), r.Desc().RaftID, index); err != nil {
-				rErr = newReplicaCorruptionError(
-					util.Errorf("could not advance applied index"), err, rErr)
-				return
+	// If we have an out of order index, there's corruption. No sense in trying
+	// to update anything or run the command. Simply return a corruption error.
+	if oldIndex := atomic.LoadUint64(&r.appliedIndex); oldIndex >= index {
+		return newReplicaCorruptionError(util.Errorf("applied index moved backwards: %d >= %d", oldIndex, index))
+	}
+
+	// Call the helper, which returns a batch containing data written
+	// during command execution and any associated error.
+	ms := engine.MVCCStats{}
+	batch, rErr := r.applyRaftCommandHelper(ctx, index, originNode, args, reply, &ms)
+	defer batch.Close()
+
+	// Advance the last applied index and commit the batch.
+	if err := setAppliedIndex(batch, r.Desc().RaftID, index); err != nil {
+		log.Fatalc(ctx, "setting applied index in a batch should never fail: %s", err)
```

Comment:
@tamird notice I changed some of these statements where we're writing to a batch to be fatal logs as a batch doesn't touch anything persistent; it's just in-memory. Failures here would not translate to any expectation of corruption and should in fact be impossible barring a bug in the code. Better to just fail the instance.